quickwit/quickwit-storage/src/timeout_and_retry_storage.rs

```diff
@@ -0,0 +1,270 @@
+use tokio::io::AsyncRead;
+
+use crate::storage::SendableAsync;
+use crate::{BulkDeleteError, PutPayload, Storage, StorageErrorKind, StorageResult};
+
+/// Storage proxy that implements a retry operation if the underlying storage
+/// takes too long.
+///
+/// This is useful in order to unsure a low latency on S3.
+/// Retrying agressively is recommended for S3.
+/// https://docs.aws.amazon.com/whitepapers/latest/s3-optimizing-performance-best-practices/timeouts-and-retries-for-latency-sensitive-applications.html
+#[derive(Clone, Debug)]
+pub struct TimeoutAndRetryStorage {
+    underlying: Arc<dyn Storage>,
+    storage_timeout_policy: StorageTimeoutPolicy,
+}
+
+impl TimeoutAndRetryStorage {
+    /// Creates a new `TimeoutAndRetryStorage`.
+    ///
+    /// See [StorageTimeoutPolicy] for more information.
+    pub fn new(storage: Arc<dyn Storage>, storage_timeout_policy: StorageTimeoutPolicy) -> Self {
+        TimeoutAndRetryStorage {
+            underlying: storage,
+            storage_timeout_policy,
+        }
+    }
+}
+
+#[async_trait]
+impl Storage for TimeoutAndRetryStorage {
+    async fn check_connectivity(&self) -> anyhow::Result<()> {
+        self.underlying.check_connectivity().await
+    }
+
+    async fn put(&self, path: &Path, payload: Box<dyn PutPayload>) -> StorageResult<()> {
+        self.underlying.put(path, payload).await
+    }
+
+    fn copy_to<'life0, 'life1, 'life2, 'async_trait>(
+        &'life0 self,
+        path: &'life1 Path,
+        output: &'life2 mut dyn SendableAsync,
+    ) -> ::core::pin::Pin<
+        Box<
+            dyn ::core::future::Future<Output = StorageResult<()>>
+                + ::core::marker::Send
+                + 'async_trait,
+        >,
+    >
+    where
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        'life2: 'async_trait,
+        Self: 'async_trait,
+    {
+        self.underlying.copy_to(path, output)
+    }
+
+    async fn copy_to_file(&self, path: &Path, output_path: &Path) -> StorageResult<u64> {
+        self.underlying.copy_to_file(path, output_path).await
+    }
+
+    /// Downloads a slice of a file from the storage, and returns an in memory buffer
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> StorageResult<OwnedBytes> {
+        let num_bytes = range.len();
+        for (attempt_id, timeout_duration) in self
+            .storage_timeout_policy
+            .compute_timeout(num_bytes)
+            .enumerate()
+        {
+            let get_slice_fut = self.underlying.get_slice(path, range.clone());
+            match tokio::time::timeout(timeout_duration, get_slice_fut).await {
+                Ok(result) => return result,
+                Err(_elapsed) => {
+                    if let Some(get_slice_timeout_count) = crate::STORAGE_METRICS
+                        .get_slice_timeout_total_by_attempts
+                        .get(attempt_id)
+                    {
+                        get_slice_timeout_count.inc();
```

Comment:
> Number of `get_slice` operations that timed out

the update of metric feels odd. Nothing is incremented if we succeed on first try (it feels like "0" should be), if we fail the first try, and succeed at 2nd, "0" is incremented (despite there be a timeout), and "1" is not. If we time out on 3rd try, "0" and "1" are incremented, but "2" isn't.
There is an off by one, and I don't think the metric should be updated on the sad path, only on the happy path, or after the last try

Reply:
yeah, it is counting timeouts and the position where they occur.
One benefit is that without filtering on the label, you get the total number of timeout.

I thought it was natural, but if you find it confusing, I can think of another solution.